feat: Node CLI wrapper and configuration tool

kevinjaypatel opened the pull request:
# Merow CLI 
Node CLI Wrapper and configuration tool to quickly spin up a running node instance. 

## Summary

As featured for the `REDACTED` hackathon, the Merow CLI tool wraps the Node CLI and provides abstract commands for setting up a Calimero Node, which can be configured using a single file.  

## Usage 

Setup nodes using the default configuration file:  `./crates/merow/config/default.toml` 

```javascript
[coordinator]
name = "coordinator" 
server_port = 2427
swarm_port = 2527
home = "data"

[admin] 
name = "node1" 
server_port = 2428 
swarm_port = 2528
home = "data"
```

### Starting up  a Coordinator (same steps apply for the client Node)
E.g. Initializes Coordinator (with defaults) 
```bash
  cargo run -p merow -- init-coordinator 
```

Start a running coordinator 
```bash
  cargo run -p merow -- start-coordinator 
```

### Accessing the Node via Admin Dashboard

- [Local Admin Dashboard](http://localhost:<coordinator.server_port>/admin-dashboard/)


## Motivation
As the Calimero Network continues to evolve, Merow CLI strives to simplify the process for Calimero Developers to create a minimal develop environment by leveraging the Calimero Node CLI.

MatejVukosav commented:
Hey @kevinjaypatel ,

Thank you for your contribution! We appreciate efforts that enhance the developer experience and make our tools more user friendly.

After reviewing your code, we believe it’s best to host it as a separate repository. This allows the core to remain focused on the fundamentals required for running nodes while keeping additional features modular.

With our recent updates, we’ve replaced the need for a coordinator with CRDTs, enabling peers to synchronize state directly. As a result, some aspects of the existing implementation may no longer be necessary.

We’d love to highlight `merow` in our documentation as an alternative setup option once it’s updated to align with these changes. Let us know if you’d like any assistance or further guidance.

Thanks again for your work!

kevinjaypatel commented:
Hi Matej,

Ah, this is great! Thanks so much for the opportunity! I am glad to
contribute, Calimero is the first open-source project I feel a part of, and
really appreciate being able to make a difference on the network! Thank you
for sharing your plans with me, and the insight for the decisions made.

I am currently updating merow to account for the changes in the
"coordinator" dependency. I am looking for a way to actively sync updates
to the forked calimero repo where merow exists, so users who are interested
can still run the latest Calimero Network. I will be sure to inform you
when that is completed, I would love to have the opportunity to share
"merow" on the developer documentation!

Thanks so much, Matej!

Best regards,
Kevin

On Wed, Nov 27, 2024 at 4:13 AM Matej Vukosav ***@***.***>
wrote:

> Hey @kevinjaypatel <https://github.com/kevinjaypatel> ,
>
> Thank you for your contribution! We appreciate efforts that enhance the
> developer experience and make our tools more user friendly.
>
> After reviewing your code, we believe it’s best to host it as a separate
> repository. This allows the core to remain focused on the fundamentals
> required for running nodes while keeping additional features modular.
>
> With our recent updates, we’ve replaced the need for a coordinator with
> CRDTs, enabling peers to synchronize state directly. As a result, some
> aspects of the existing implementation may no longer be necessary.
>
> We’d love to highlight merow in our documentation as an alternative setup
> option once it’s updated to align with these changes. Let us know if you’d
> like any assistance or further guidance.
>
> Thanks again for your work!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/calimero-network/core/pull/959#issuecomment-2503720132>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKIUQKFF3NBAVSWFP4QUO3T2CWZNVAVCNFSM6AAAAABROVRTMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMBTG4ZDAMJTGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kevinjaypatel commented:
Dear Matej,

I updated the "Merow" repository. Can you please provide any feedback you
may have? I would love for it to be featured as an alternative for
developers to use when setting up the Calimero Network.

You can the most recent version here: https://github.com/kevinjaypatel/core

Best regards,
Kevin

On Mon, Dec 2, 2024 at 11:26 AM Kevin Patel ***@***.***> wrote:

> Hi Matej,
>
> Ah, this is great! Thanks so much for the opportunity! I am glad to
> contribute, Calimero is the first open-source project I feel a part of, and
> really appreciate being able to make a difference on the network! Thank you
> for sharing your plans with me, and the insight for the decisions made.
>
> I am currently updating merow to account for the changes in the
> "coordinator" dependency. I am looking for a way to actively sync updates
> to the forked calimero repo where merow exists, so users who are interested
> can still run the latest Calimero Network. I will be sure to inform you
> when that is completed, I would love to have the opportunity to share
> "merow" on the developer documentation!
>
> Thanks so much, Matej!
>
> Best regards,
> Kevin
>
> On Wed, Nov 27, 2024 at 4:13 AM Matej Vukosav ***@***.***>
> wrote:
>
>> Hey @kevinjaypatel <https://github.com/kevinjaypatel> ,
>>
>> Thank you for your contribution! We appreciate efforts that enhance the
>> developer experience and make our tools more user friendly.
>>
>> After reviewing your code, we believe it’s best to host it as a separate
>> repository. This allows the core to remain focused on the fundamentals
>> required for running nodes while keeping additional features modular.
>>
>> With our recent updates, we’ve replaced the need for a coordinator with
>> CRDTs, enabling peers to synchronize state directly. As a result, some
>> aspects of the existing implementation may no longer be necessary.
>>
>> We’d love to highlight merow in our documentation as an alternative
>> setup option once it’s updated to align with these changes. Let us know if
>> you’d like any assistance or further guidance.
>>
>> Thanks again for your work!
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/calimero-network/core/pull/959#issuecomment-2503720132>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AKIUQKFF3NBAVSWFP4QUO3T2CWZNVAVCNFSM6AAAAABROVRTMWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMBTG4ZDAMJTGI>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

MatejVukosav commented:
Sure, just checked the code and have few comments.

1. can u call repo `merow`?
2. you don't need core code. You can remove it and just have `merow` implementation
